pylint/testutils/functional/find_functional_tests.py

```diff
@@ -49,39 +51,89 @@ def get_functional_test_files_from_directory(
     return suite
 
 
-def _check_functional_tests_structure(directory: Path) -> None:
-    """Check if test directories follow correct file/folder structure."""
-    # Ignore underscored directories
+def _check_functional_tests_structure(
+    directory: Path, max_file_per_directory: int
+) -> None:
+    """Check if test directories follow correct file/folder structure.
+
+    Ignore underscored directories or files.
+    """
     if Path(directory).stem.startswith("_"):
         return
 
     files: set[Path] = set()
     dirs: set[Path] = set()
 
+    def _get_files_from_dir(
+        path: Path, violations: list[tuple[Path, int]]
+    ) -> list[Path]:
+        """Return directories and files from a directory and handles violations."""
+        files_without_leading_underscore = list(
+            p for p in path.iterdir() if not str(p).startswith("_")
+        )
+        if len(files_without_leading_underscore) > max_file_per_directory:
+            violations.append((path, len(files_without_leading_underscore)))
+        return files_without_leading_underscore
+
+    def walk(path: Path) -> Iterator[Path]:
+        violations: list[tuple[Path, int]] = []
+        violations_msgs: set[str] = set()
+        parent_dir_files = _get_files_from_dir(path, violations)
+        error_msg = (
```

Comment:
Let's just assign this on L98. I don't think there is a need to do so here?

Reply:
Feel like a nitpick, we have a lot to do in pylint, and this is blocking 3.0.0a6, moving assignments around in functional test's tests is not an efficient use of our time.

Reply:
We need it line 93 in ``violations_msgs.add(str(e).replace(error_msg, ""))``

Reply:
That I completely missed! Resolving!

Reply:
I understand that you feel this way but I must say that only through that scrutiny we found the logic flaw in `violations` that would have actually added a regression in this PR that is supposed to only add coverage.

There is a fine balance to strike between scrutiny and just merging because we want to continue and perhaps in this case I was a bit too strict, but I do think we should be somewhat strict about our code. Stuff like the adding of default values to `__init__` and `postinit` in `astroid` and some of the non-understandable tests we have there could all have been avoided by carefully considering the PRs and the code in them.

As you know I do really appreciate the effort that went into this PR and I know this was hard to refactor and add tests for. I have only one open comment and then we can go ahead and merge this 🚀

Reply:
It's true that reviewing made us realize a problem but even if we merged with the violations msg issue, it's an error in a helper that enforce that there's not too much files in our functional test directory. What I mean is that we could live with that kind of issues 😄 Especially when we're living just fine with the multiprocessing that give very wrong results when activated (without even being that much faster), or #3512 not implemented 3 years after being opened, the new json reporter either, astroid not typed, or even typable, etc. etc. you know as well as me that some issues are pretty important and pretty much only doable by a maintainer because they are so hard to design or start on. I think that sometime when the cost of a mistake is low we should *just* merge because the changes makes the code better even if it's not perfect and because we're going to loose contributors if they dread our reviews.